tools/shadowsim/src/shadowsim/shadow_exec.py

```diff
@@ -0,0 +1,249 @@
+                simulation config and results.
+    stdout -- Destination for the simulated program's merged stdout and stderr.
+    stderr -- Destination for other "meta" output.
+    shadow_bin -- Shadow binary basename or path.
+    model_unblocked_syscall_latency -- Whether to set shadow's --model-unblocked-syscall-latency.
+    """
+
+    tmpdir = Path(tempfile.mkdtemp(prefix=f"{progname}-", dir=temp_dir))
+
+    # quote each argument individually
+    arg_str = " ".join(shlex.quote(a) for a in args)
+
+    wrapper_script = textwrap.dedent(
+        f"""
+        set -euo pipefail
+
+        # Change back to host working dir
+        cd {shlex.quote(str(Path('.').resolve()))}
+
+        # Run specified command, merging stderr to stdout
+        exec {arg_str} 2>&1
+        """
+    )
+
+    config = scfg.Config(
+        general=scfg.General(
+            # Experimentally, this is about the highest we can go before
+            # the time spent fast forwarding empty time gets noticeable.
+            stop_time="1000000h",
+            log_level="warning",
+            heartbeat_interval=f"{2**63}ns",
+            model_unblocked_syscall_latency=model_unblocked_syscall_latency,
+        ),
+        network=scfg.Network(graph=scfg.Graph(type="1_gbit_switch")),
+        hosts={
+            "host": scfg.Host(
+                network_node_id=0,
+                processes=[
+                    scfg.Process(
+                        path="bash",
+                        args=[
+                            "-c",
+                            wrapper_script,
+                        ],
+                    )
+                ],
+            )
+        },
+    )
+    config_path = tmpdir.joinpath("shadow.yaml")
+    config_path.write_text(yaml.safe_dump(config))
+
+    data_dir = tmpdir.joinpath("shadow.data")
+    shadow_args = [
+        f"--data-directory={data_dir}",
+        "--progress=false",
+        str(config_path),
+    ]
+
+    shadow_log_path = tmpdir.joinpath("shadow.log")
+    with shadow_log_path.open("w") as shadow_log:
+        shadow_ps = subprocess.Popen(
+            [str(shadow_bin)] + shadow_args,
+            stdout=shadow_log,
+            stderr=subprocess.PIPE,
+        )
+        shadow_stdout = None
+        shadow_exited = False
+        while True:
+            # Try opening the simulated process's stdout if we haven't
+            # successfully done so yet.
+            if shadow_stdout is None:
+                try:
+                    shadow_stdout = open(
+                        str(data_dir.joinpath("hosts/host/bash.1000.stdout")), "rb"
+                    )
+                except FileNotFoundError:
+                    # Not created yet, presumably
+                    pass
+
+            # Pump data from sim stdout to our stdout
+            data = None
+            if shadow_stdout is not None:
+                data = shadow_stdout.read()
+            if data:
+                stdout.write(data)
```

Comment:
The docs for `write` say "Write the string s to the stream and return the number of characters written". It sounds like this isn't guaranteed to write all bytes? Does this need to be called in a loop so that data isn't lost? Or maybe just assert for now that the expected number was written.

Reply:
Oof good catch. I'd assumed it always writes all of the data since `read` does so (without an explicit limit set). [`RawIOBase.write`](https://docs.python.org/3/library/io.html#io.RawIOBase.write) is a bit more explicit that the write can be partial. `TextIOBase.write` might or might not try harder; as you point out there's no promise.

Added a loop.